components/core/JsInterop/DomEventListener.cs

```diff
@@ -0,0 +1,225 @@
+        private Dictionary<string, IDisposable> _dotNetObjectStore = new();
+        private bool? _isResizeObserverSupported = null;
+
+        private readonly IJSRuntime _jsRuntime;
+        private readonly DomEventSubscriptionStore _domEventSubscriptionsStore;
+
+        public DomEventListener(IJSRuntime jsRuntime, DomEventSubscriptionStore domEventSubscriptionStore)
+        {
+            _jsRuntime = jsRuntime;
+            _domEventSubscriptionsStore = domEventSubscriptionStore;
+        }
+
+        private static string FormatKey(object dom, string eventName) => $"{dom}-{eventName}";
+
+        public void AddExclusive<T>(object dom, string eventName, Action<T> callback, bool preventDefault = false)
+        {
+            var key = FormatKey(dom, eventName);
+            if (_dotNetObjectStore.ContainsKey(key))
+                return;
+
+            var dotNetObject = DotNetObjectReference.Create(new Invoker<T>((p) =>
+            {
+                callback(p);
+            }));
+            _jsRuntime.InvokeAsync<string>(JSInteropConstants.AddDomEventListener, dom, eventName, preventDefault, dotNetObject);
+            _dotNetObjectStore.Add(key, dotNetObject);
+        }
+
+        public void RemoveExclusive(object dom, string eventName)
+        {
+            var key = FormatKey(dom, eventName);
+            if (_dotNetObjectStore.TryGetValue(key, out IDisposable value))
+            {
+                value.Dispose();
+            }
+            _dotNetObjectStore.Remove(key);
+        }
+
+        public void DisposeExclusive()
+        {
+            foreach (var (k, v) in _dotNetObjectStore)
+            {
+                v.Dispose();
+            }
+            _dotNetObjectStore.Clear();
+        }
+
+        #region SharedEventListerner
+
+        public virtual void AddShared<T>(object dom, string eventName, Action<T> callback, bool preventDefault = false)
+        {
+            string key = FormatKey(dom, eventName);
+            if (!_domEventSubscriptionsStore.ContainsKey(key))
+            {
+                _domEventSubscriptionsStore[key] = new List<DomEventSubscription>();
+
+                var dotNetObject = DotNetObjectReference.Create(new Invoker<string>((p) =>
+                {
+                    for (var i = 0; i < _domEventSubscriptionsStore[key].Count; i++)
+                    {
+                        var subscription = _domEventSubscriptionsStore[key][i];
+                        object tP = JsonSerializer.Deserialize(p, subscription.Type);
+                        subscription.Delegate.DynamicInvoke(tP);
+                    }
+                }));
+
+                _jsRuntime.InvokeAsync<string>(JSInteropConstants.AddDomEventListener, dom, eventName, preventDefault, dotNetObject);
+            }
+            _domEventSubscriptionsStore[key].Add(new DomEventSubscription(callback, typeof(T)));
+        }
+
+        public void RemoveShared<T>(object dom, string eventName, Action<T> callback)
```

Comment:
Can the RemoveShared method be used without a generic type? That would be more convenient.

Reply:
Change `Action<T>` to Delegate

Reply:
![image](https://user-images.githubusercontent.com/4554555/132160603-7e9c8a28-1500-4daa-a714-5fe4963780bd.png)

![image](https://user-images.githubusercontent.com/4554555/132160623-d487bb47-ffb6-4815-84ce-d5033fc5075c.png)

I have encountered the error above after changed to Delegate. Any suggestion? I do not quite understand how the Delegate work.

Reply:
@ElderJames , added DisposeShared method to remove all the callback in a single statement

Reply:
Cool!